[Bug]: Panic while checking backup container ID if pod container statuses are missing

### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Contact Details

_No response_

### Version

1.23.1

### What version of Kubernetes are you using?

1.30 (unsupported)

### What is your Kubernetes environment?

Self-managed: kind (evaluation)

### How did you install the operator?

YAML manifest

### What happened?

{"level":"info","ts":"2024-05-28T07:09:22Z","msg":"Observed a panic in reconciler: runtime error: index out of range [0] with length 0","controller":"backup","controllerGroup":"postgresql.k8s.enterprisedb.io","controllerKind":"Backup","Backup":{"name":"p-r55dk9edcv-sch-20240515063800","namespace":"default"},"namespace":"default","name":"p-r55dk9edcv-sch-20240515063800","reconcileID":"62369a3b-fc02-48b2-8245-91d78ee3b571"}

panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

### Cluster resource

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/api/v1/backup_types.go
+++ b/api/v1/backup_types.go
@@ -328,14 +328,10 @@ func (backupStatus *BackupStatus) SetAsCompleted() {
 // SetAsStarted marks a certain backup as started
 func (backupStatus *BackupStatus) SetAsStarted(targetPod *corev1.Pod, method BackupMethod) {
 	backupStatus.Phase = BackupPhaseStarted
-
-	if len(targetPod.Status.ContainerStatuses) > 0 {
-		backupStatus.InstanceID = &InstanceID{
-			PodName:     targetPod.Name,
-			ContainerID: targetPod.Status.ContainerStatuses[0].ContainerID,
-		}
+	backupStatus.InstanceID = &InstanceID{
+		PodName:     targetPod.Name,
+		ContainerID: targetPod.Status.ContainerStatuses[0].ContainerID,
 	}
-
 	backupStatus.Method = method
 }
 

--- a/api/v1/backup_types.go
+++ b/api/v1/backup_types.go
@@ -328,10 +328,14 @@ func (backupStatus *BackupStatus) SetAsCompleted() {
 // SetAsStarted marks a certain backup as started
 func (backupStatus *BackupStatus) SetAsStarted(targetPod *corev1.Pod, method BackupMethod) {
 	backupStatus.Phase = BackupPhaseStarted
-	backupStatus.InstanceID = &InstanceID{
-		PodName:     targetPod.Name,
-		ContainerID: targetPod.Status.ContainerStatuses[0].ContainerID,
+
+	if len(targetPod.Status.ContainerStatuses) > 0 {
+		backupStatus.InstanceID = &InstanceID{
+			PodName:     targetPod.Name,
+			ContainerID: targetPod.Status.ContainerStatuses[0].ContainerID,
+		}
 	}
+
 	backupStatus.Method = method
 }
 

--- a/api/v1/backup_types.go
+++ b/api/v1/backup_types.go
@@ -326,11 +326,11 @@ func (backupStatus *BackupStatus) SetAsCompleted() {
 }
 
 // SetAsStarted marks a certain backup as started
-func (backupStatus *BackupStatus) SetAsStarted(targetPod *corev1.Pod, method BackupMethod) {
+func (backupStatus *BackupStatus) SetAsStarted(podName, containerID string, method BackupMethod) {
 	backupStatus.Phase = BackupPhaseStarted
 	backupStatus.InstanceID = &InstanceID{
-		PodName:     targetPod.Name,
-		ContainerID: targetPod.Status.ContainerStatuses[0].ContainerID,
+		PodName:     podName,
+		ContainerID: containerID,
 	}
 	backupStatus.Method = method
 }

--- a/api/v1/backup_types_test.go
+++ b/api/v1/backup_types_test.go
@@ -46,7 +46,7 @@ var _ = Describe("BackupStatus structure", func() {
 			},
 		}
 
-		status.SetAsStarted(&pod, BackupMethodBarmanObjectStore)
+		status.SetAsStarted(pod.Name, pod.Status.ContainerStatuses[0].ContainerID, BackupMethodBarmanObjectStore)
 		Expect(status.Phase).To(BeEquivalentTo(BackupPhaseStarted))
 		Expect(status.InstanceID).ToNot(BeNil())
 		Expect(status.InstanceID.PodName).To(Equal("cluster-example-1"))

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -197,7 +197,7 @@ func (r *BackupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		}
 		contextLogger.Debug("Found pod for backup", "pod", pod.Name)
 
-		if !utils.IsPodReady(*pod) || !utils.PodHasContainerStatuses(*pod) {
+		if !utils.IsPodReady(*pod) {
 			contextLogger.Info("Backup target is not ready, will retry in 30 seconds", "target", pod.Name)
 			backup.Status.Phase = apiv1.BackupPhasePending
 			r.Recorder.Eventf(&backup, "Warning", "BackupPending", "Backup target pod not ready: %s",
@@ -288,7 +288,7 @@ func (r *BackupReconciler) isValidBackupRunning(
 	}
 
 	var containerIsNotRestarted bool
-	if utils.PodHasContainerStatuses(pod) && backup.Status.InstanceID != nil {
+	if utils.PodHasContainerStatuses(pod) {
 		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	}
 	isPodActive := utils.IsPodActive(pod)
@@ -374,7 +374,11 @@ func (r *BackupReconciler) reconcileSnapshotBackup(
 	}
 
 	if len(backup.Status.Phase) == 0 || backup.Status.Phase == apiv1.BackupPhasePending {
-		backup.Status.SetAsStarted(targetPod, apiv1.BackupMethodVolumeSnapshot)
+		backup.Status.SetAsStarted(
+			targetPod.Name,
+			targetPod.Status.ContainerStatuses[0].ContainerID,
+			apiv1.BackupMethodVolumeSnapshot,
+		)
 		// given that we use only kubernetes resources we can use the backup name as ID
 		backup.Status.BackupID = backup.Name
 		backup.Status.BackupName = backup.Name
@@ -569,7 +573,7 @@ func startInstanceManagerBackup(
 ) error {
 	// This backup has been started
 	status := backup.GetStatus()
-	status.SetAsStarted(pod, backup.Spec.Method)
+	status.SetAsStarted(pod.Name, pod.Status.ContainerStatuses[0].ContainerID, backup.Spec.Method)
 
 	if err := postgres.PatchBackupStatusAndRetry(ctx, client, backup); err != nil {
 		return err

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -288,7 +288,7 @@ func (r *BackupReconciler) isValidBackupRunning(
 	}
 
 	var containerIsNotRestarted bool
-	if !utils.IsPodReady(pod) && backup.Status.InstanceID != nil {
+	if utils.PodHasContainerStatuses(pod) && backup.Status.InstanceID != nil {
 		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	}
 	isPodActive := utils.IsPodActive(pod)

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -288,7 +288,7 @@ func (r *BackupReconciler) isValidBackupRunning(
 	}
 
 	var containerIsNotRestarted bool
-	if len(pod.Status.ContainerStatuses) > 0 && backup.Status.InstanceID != nil {
+	if !utils.IsPodReady(pod) && backup.Status.InstanceID != nil {
 		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	}
 	isPodActive := utils.IsPodActive(pod)

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -287,7 +287,10 @@ func (r *BackupReconciler) isValidBackupRunning(
 		return false, fmt.Errorf("unknown.spec.target received: %s", backup.Spec.Target)
 	}
 
-	containerIsNotRestarted := backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
+	var containerIsNotRestarted bool
+	if len(pod.Status.ContainerStatuses) > 0 {
+		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
+	}
 	isPodActive := utils.IsPodActive(pod)
 	if isCorrectPodElected && containerIsNotRestarted && isPodActive {
 		contextLogger.Info("Backup is already running on",

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -197,7 +197,7 @@ func (r *BackupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		}
 		contextLogger.Debug("Found pod for backup", "pod", pod.Name)
 
-		if !utils.IsPodReady(*pod) {
+		if !utils.IsPodReady(*pod) || !utils.PodHasContainerStatuses(*pod) {
 			contextLogger.Info("Backup target is not ready, will retry in 30 seconds", "target", pod.Name)
 			backup.Status.Phase = apiv1.BackupPhasePending
 			r.Recorder.Eventf(&backup, "Warning", "BackupPending", "Backup target pod not ready: %s",
@@ -209,7 +209,7 @@ func (r *BackupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 			"cluster", cluster.Name,
 			"pod", pod.Name)
 
-		// This backup has been started
+		// This backup can be started
 		if err := startInstanceManagerBackup(ctx, r.Client, &backup, pod, &cluster); err != nil {
 			r.Recorder.Eventf(&backup, "Warning", "Error", "Backup exit with error %v", err)
 			tryFlagBackupAsFailed(ctx, r.Client, &backup, fmt.Errorf("encountered an error while taking the backup: %w", err))
@@ -369,8 +369,8 @@ func (r *BackupReconciler) reconcileSnapshotBackup(
 		return &ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 	}
 
-	if len(targetPod.Status.ContainerStatuses) == 0 {
-		return nil, fmt.Errorf("pod container statuses list is empty")
+	if !utils.PodHasContainerStatuses(*targetPod) {
+		return nil, fmt.Errorf("target pod lacks container statuses")
 	}
 
 	if len(backup.Status.Phase) == 0 || backup.Status.Phase == apiv1.BackupPhasePending {
@@ -567,10 +567,6 @@ func startInstanceManagerBackup(
 	pod *corev1.Pod,
 	cluster *apiv1.Cluster,
 ) error {
-	if len(pod.Status.ContainerStatuses) == 0 {
-		return fmt.Errorf("pod container statuses list is empty")
-	}
-
 	// This backup has been started
 	status := backup.GetStatus()
 	status.SetAsStarted(pod, backup.Spec.Method)

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -288,7 +288,7 @@ func (r *BackupReconciler) isValidBackupRunning(
 	}
 
 	var containerIsNotRestarted bool
-	if len(pod.Status.ContainerStatuses) > 0 {
+	if len(pod.Status.ContainerStatuses) > 0 && backup.Status.InstanceID != nil {
 		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	}
 	isPodActive := utils.IsPodActive(pod)

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -287,10 +287,7 @@ func (r *BackupReconciler) isValidBackupRunning(
 		return false, fmt.Errorf("unknown.spec.target received: %s", backup.Spec.Target)
 	}
 
-	var containerIsNotRestarted bool
-	if utils.PodHasContainerStatuses(pod) {
-		containerIsNotRestarted = backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
-	}
+	containerIsNotRestarted := utils.PodHasContainerStatuses(pod) && backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	isPodActive := utils.IsPodActive(pod)
 	if isCorrectPodElected && containerIsNotRestarted && isPodActive {
 		contextLogger.Info("Backup is already running on",

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -369,6 +369,10 @@ func (r *BackupReconciler) reconcileSnapshotBackup(
 		return &ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 	}
 
+	if len(targetPod.Status.ContainerStatuses) == 0 {
+		return nil, fmt.Errorf("pod container statuses list is empty")
+	}
+
 	if len(backup.Status.Phase) == 0 || backup.Status.Phase == apiv1.BackupPhasePending {
 		backup.Status.SetAsStarted(targetPod, apiv1.BackupMethodVolumeSnapshot)
 		// given that we use only kubernetes resources we can use the backup name as ID
@@ -563,6 +567,10 @@ func startInstanceManagerBackup(
 	pod *corev1.Pod,
 	cluster *apiv1.Cluster,
 ) error {
+	if len(pod.Status.ContainerStatuses) == 0 {
+		return fmt.Errorf("pod container statuses list is empty")
+	}
+
 	// This backup has been started
 	status := backup.GetStatus()
 	status.SetAsStarted(pod, backup.Spec.Method)

--- a/internal/controller/backup_controller.go
+++ b/internal/controller/backup_controller.go
@@ -287,7 +287,8 @@ func (r *BackupReconciler) isValidBackupRunning(
 		return false, fmt.Errorf("unknown.spec.target received: %s", backup.Spec.Target)
 	}
 
-	containerIsNotRestarted := utils.PodHasContainerStatuses(pod) && backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
+	containerIsNotRestarted := utils.PodHasContainerStatuses(pod) &&
+		backup.Status.InstanceID.ContainerID == pod.Status.ContainerStatuses[0].ContainerID
 	isPodActive := utils.IsPodActive(pod)
 	if isCorrectPodElected && containerIsNotRestarted && isPodActive {
 		contextLogger.Info("Backup is already running on",

--- a/pkg/utils/pod_conditions.go
+++ b/pkg/utils/pod_conditions.go
@@ -55,6 +55,11 @@ func IsPodReady(pod corev1.Pod) bool {
 	return false
 }
 
+// PodHasContainerStatuses checks if a Pod has container status elements
+func PodHasContainerStatuses(pod corev1.Pod) bool {
+	return len(pod.Status.ContainerStatuses) > 0
+}
+
 // IsPodActive checks if a pod is active, copied from:
 // https://github.com/kubernetes/kubernetes/blob/1bd0077/test/e2e/framework/pod/resource.go#L664
 func IsPodActive(p corev1.Pod) bool {

--- a/tests/e2e/upgrade_test.go
+++ b/tests/e2e/upgrade_test.go
@@ -535,7 +535,9 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
 			podList, err := env.GetClusterPodList(upgradeNamespace, clusterName1)
 			Expect(err).ToNot(HaveOccurred())
 			for _, pod := range podList.Items {
-				Expect(pod.Status.ContainerStatuses[0].RestartCount).To(BeEquivalentTo(0))
+				if len(pod.Status.ContainerStatuses) > 0 {
+					Expect(pod.Status.ContainerStatuses[0].RestartCount).To(BeEquivalentTo(0))
+				}
 			}
 		})
 

--- a/tests/e2e/upgrade_test.go
+++ b/tests/e2e/upgrade_test.go
@@ -535,9 +535,7 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
 			podList, err := env.GetClusterPodList(upgradeNamespace, clusterName1)
 			Expect(err).ToNot(HaveOccurred())
 			for _, pod := range podList.Items {
-				if len(pod.Status.ContainerStatuses) == 0 {
-					Fail("Pod container statuses list is empty")
-				}
+				Expect(pod.Status.ContainerStatuses).NotTo(BeEmpty())
 				Expect(pod.Status.ContainerStatuses[0].RestartCount).To(BeEquivalentTo(0))
 			}
 		})

--- a/tests/e2e/upgrade_test.go
+++ b/tests/e2e/upgrade_test.go
@@ -535,9 +535,10 @@ var _ = Describe("Upgrade", Label(tests.LabelUpgrade, tests.LabelNoOpenshift), O
 			podList, err := env.GetClusterPodList(upgradeNamespace, clusterName1)
 			Expect(err).ToNot(HaveOccurred())
 			for _, pod := range podList.Items {
-				if len(pod.Status.ContainerStatuses) > 0 {
-					Expect(pod.Status.ContainerStatuses[0].RestartCount).To(BeEquivalentTo(0))
+				if len(pod.Status.ContainerStatuses) == 0 {
+					Fail("Pod container statuses list is empty")
 				}
+				Expect(pod.Status.ContainerStatuses[0].RestartCount).To(BeEquivalentTo(0))
 			}
 		})
 

